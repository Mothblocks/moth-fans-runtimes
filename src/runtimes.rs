use std::path::PathBuf;

use color_eyre::eyre::{Context, ContextCompat};
use regex::Regex;
use serde::{Deserialize, Serialize};

const RUNTIME_REGEX_PATTERN: &str = r#"The following runtime has occurred (?P<count>[0-9]+).*
runtime error: (?P<exception>.+)
proc name: (?P<proc>.+?) \((?P<proc_path>.+?)\)
  source file: (?P<source_file>.+?),(?P<line>[0-9]+)"#;

#[derive(Deserialize, Serialize)]
pub struct RuntimeBatch {
    pub count: u64,
    pub exception: String,
    pub proc_path: String,
    pub source_file: String,
    pub line: u64,

    pub best_guess_filenames: Option<BestGuessFilenames>,
}

#[derive(Deserialize, Serialize)]
pub enum BestGuessFilenames {
    Definitely(PathBuf),
    Possible(Vec<PathBuf>),
}

impl RuntimeBatch {
    fn patch_special_procs(&mut self) {
        if self.proc_path == "/proc/_stack_trace" {
            self.patch_stack_trace();
        }
    }

    fn patch_stack_trace(&mut self) {
        let stack_trace_regex = Regex::new(r".+\(((?P<filename>.+?):(?P<line>[0-9]+))\)$").unwrap();

        if let Some(captures) = stack_trace_regex.captures(&self.exception) {
            let filename = PathBuf::from(captures.name("filename").unwrap().as_str());

            self.source_file = filename.file_name().unwrap().to_string_lossy().to_string();
            self.best_guess_filenames = Some(BestGuessFilenames::Definitely(filename));

            self.line = captures.name("line").unwrap().as_str().parse().unwrap();
        }
    }
}

fn group_to_runtime_batch(captures: &regex::Captures) -> color_eyre::Result<RuntimeBatch> {
    let count = captures
        .name("count")
        .context("error getting count")?
        .as_str()
        .parse::<u64>()
        .context("error parsing count")?;

    let exception = captures
        .name("exception")
        .context("error getting exception")?
        .as_str()
        .to_string();

    let proc_path = captures
        .name("proc_path")
        .context("error getting proc_path")?
        .as_str()
        .to_string();

    let source_file = captures
        .name("source_file")
        .context("error getting source_file")?
        .as_str()
        .to_string();

    let line = captures
        .name("line")
        .context("error getting line")?
        .as_str()
        .parse::<u64>()
        .context("error parsing line")?;

    let mut runtime_batch = RuntimeBatch {
        count,
        exception,
        proc_path,
        source_file,
        line,
        best_guess_filenames: None,
    };

    runtime_batch.patch_special_procs();

    Ok(runtime_batch)
}

pub fn get_runtimes_for_round(
    runtime_condensed_txt: &str,
) -> color_eyre::Result<Vec<RuntimeBatch>> {
    let runtime_regex = Regex::new(RUNTIME_REGEX_PATTERN).unwrap();

    runtime_regex
        .captures_iter(runtime_condensed_txt)
        .map(|captures| -> color_eyre::Result<RuntimeBatch> {
            match group_to_runtime_batch(&captures) {
                Ok(runtime_batch) => Ok(runtime_batch),
                result @ Err(_) => result.context(format!(
                    "error while parsing {}",
                    captures.get(0).expect("error getting group 0").as_str()
                )),
            }
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_runtimes_for_round_191838() {
        insta::assert_json_snapshot!(get_runtimes_for_round(include_str!(
            "./test_data/191838-runtime.condensed.txt"
        ))
        .unwrap());
    }
}
