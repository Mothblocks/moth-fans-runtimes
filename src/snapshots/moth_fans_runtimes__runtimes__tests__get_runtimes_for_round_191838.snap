---
source: src/runtimes.rs
assertion_line: 132
expression: "get_runtimes_for_round(include_str!(\"./test_data/191838-runtime.condensed.txt\")).unwrap()"
---
[
  {
    "count": 21886,
    "exception": "Cannot execute null.return air().",
    "proc_path": "/obj/machinery/meter/process_atmos",
    "source_file": "meter.dm",
    "line": 58,
    "best_guess_filename": null
  },
  {
    "count": 511,
    "exception": "Cannot execute null.archive().",
    "proc_path": "/turf/open/process_cell",
    "source_file": "LINDA_turf_tile.dm",
    "line": 281,
    "best_guess_filename": null
  },
  {
    "count": 240,
    "exception": "undefined variable /turf/closed/mineral/var/run_later",
    "proc_path": "/turf/open/process_cell",
    "source_file": "LINDA_turf_tile.dm",
    "line": 276,
    "best_guess_filename": null
  },
  {
    "count": 53,
    "exception": "undefined variable /turf/closed/wall/var/run_later",
    "proc_path": "/turf/open/process_cell",
    "source_file": "LINDA_turf_tile.dm",
    "line": 276,
    "best_guess_filename": null
  },
  {
    "count": 31,
    "exception": "undefined variable /turf/closed/wall/mineral/iron/var/run_later",
    "proc_path": "/turf/open/process_cell",
    "source_file": "LINDA_turf_tile.dm",
    "line": 276,
    "best_guess_filename": null
  },
  {
    "count": 16,
    "exception": "Cannot read null.len",
    "proc_path": "/world/proc/preloader_setup",
    "source_file": "preloader.dm",
    "line": 12,
    "best_guess_filename": null
  },
  {
    "count": 16,
    "exception": "_queue_verb() returned false because it was given an invalid callback! (code/controllers/subsystem/verb_manager.dm:61)",
    "proc_path": "/proc/_stack_trace",
    "source_file": "verb_manager.dm",
    "line": 61,
    "best_guess_filename": ") returned false because it was given an invalid callback! (code/controllers/subsystem/verb_manager.dm"
  },
  {
    "count": 16,
    "exception": "Cannot create objects of type null.",
    "proc_path": "/datum/parsed_map/proc/build_coordinate",
    "source_file": "reader.dm",
    "line": 472,
    "best_guess_filename": null
  },
  {
    "count": 11,
    "exception": "Cannot execute null.react().",
    "proc_path": "/obj/machinery/atmospherics/components/tank/process_atmos",
    "source_file": "tank.dm",
    "line": 144,
    "best_guess_filename": null
  },
  {
    "count": 7,
    "exception": "Cannot execute null.reset cooldowns().",
    "proc_path": "/turf/open/process_cell",
    "source_file": "LINDA_turf_tile.dm",
    "line": 346,
    "best_guess_filename": null
  },
  {
    "count": 3,
    "exception": "undefined variable /turf/closed/wall/r_wall/var/run_later",
    "proc_path": "/turf/open/process_cell",
    "source_file": "LINDA_turf_tile.dm",
    "line": 276,
    "best_guess_filename": null
  },
  {
    "count": 2,
    "exception": "Cannot read null.eye",
    "proc_path": "/datum/hud/proc/update_parallax",
    "source_file": "parallax.dm",
    "line": 172,
    "best_guess_filename": null
  },
  {
    "count": 2,
    "exception": "Cannot read null.parallax_layers_cached",
    "proc_path": "/datum/hud/proc/remove_parallax",
    "source_file": "parallax.dm",
    "line": 41,
    "best_guess_filename": null
  },
  {
    "count": 2,
    "exception": "Cannot modify null.parallax_layers_max.",
    "proc_path": "/datum/hud/proc/apply_parallax_pref",
    "source_file": "parallax.dm",
    "line": 68,
    "best_guess_filename": null
  },
  {
    "count": 1,
    "exception": "Cannot read null.client",
    "proc_path": "/datum/action/vote/Remove",
    "source_file": "vote.dm",
    "line": 332,
    "best_guess_filename": null
  },
  {
    "count": 1,
    "exception": "Cannot execute null.resolve().",
    "proc_path": "/datum/forensics/proc/check_blood",
    "source_file": "_forensics.dm",
    "line": 232,
    "best_guess_filename": null
  },
  {
    "count": 1,
    "exception": "Cannot read null.x",
    "proc_path": "/atom/movable/screen/map_view/gateway_port/proc/update_portal_filters",
    "source_file": "gateway.dm",
    "line": 434,
    "best_guess_filename": null
  },
  {
    "count": 1,
    "exception": "Cannot read null.visual_delay",
    "proc_path": "/datum/move_loop/process",
    "source_file": "movement_types.dm",
    "line": 111,
    "best_guess_filename": null
  },
  {
    "count": 1,
    "exception": "undefined variable /mob/living/simple_animal/mouse/var/w_uniform",
    "proc_path": "/datum/hallucination/fake_item/start",
    "source_file": "inhand_fake_item.dm",
    "line": 19,
    "best_guess_filename": null
  },
  {
    "count": 1,
    "exception": "Cannot read null.prefs",
    "proc_path": "/proc/tgui_input_list",
    "source_file": "list.dm",
    "line": 25,
    "best_guess_filename": null
  },
  {
    "count": 1,
    "exception": "addtimer called with a callback assigned to a qdeleted object. In the future such timers will not be supported and may refuse to run or run with a 0 wait (code/controllers/subsystem/timer.dm:583)",
    "proc_path": "/proc/_stack_trace",
    "source_file": "timer.dm",
    "line": 583,
    "best_guess_filename": "code/controllers/subsystem/timer.dm"
  }
]
